### ISSUE TITLE
EOF during download

### ISSUE THREAD

**Issue by scottrus** [jfrog/jfrog-cli-go#232]
We are running with up to 6 retires but still see an occasional EOF during download. Using jfrog version 1.19.1. I'm not finding anything in the Artifactory server logs at the time of this event but I'm also looking for a needle in a haystack. 

I understand that network / connection related issues may be the most likely for this EOF error. How can we better troubleshoot it?

Note that the log is truncated to only show the download and thread that error out.

```
11:05:09 + /auto/releng/bin/jfrog rt dl --retries=6 --server-id=sjcart --threads=10 --split-count=10 'builds-local/package-dev_default/14749/*' /auto/artifacts/
11:05:09 [Info] Searching items to download...
11:05:21 [Info] [Thread 2] Downloading builds-local/package-dev_default/14749/usb_install-opt/initrd.img
11:05:28 [Info] [Thread 2] [5]: 206 Partial Content...
11:05:28 [Info] [Thread 2] [3]: 206 Partial Content...
11:05:28 [Info] [Thread 2] [0]: 206 Partial Content...
11:05:28 [Info] [Thread 2] [1]: 206 Partial Content...
11:05:28 [Info] [Thread 2] [4]: 206 Partial Content...
11:05:28 [Info] [Thread 2] [7]: 206 Partial Content...
11:05:28 [Info] [Thread 2] [6]: 206 Partial Content...
11:05:28 [Info] [Thread 2] [9]: 206 Partial Content...
11:05:28 [Info] [Thread 2] [8]: 206 Partial Content...
11:05:28 [Info] [Thread 2] [2]: 206 Partial Content...
11:05:28 [Error] [Thread 2]  Received an error: unexpected EOF
11:06:43 [Error] unexpected EOF
11:06:44 {
11:06:44   "status": "failure",
11:06:44   "totals": {
11:06:44     "success": 30,
11:06:44     "failure": 1
11:06:44   }
11:06:44 }
11:06:44 [Error] Download finished with errors. Please review the logs
```

**Comment by scottrus** [comment]
IN the code logic, does the `--retires` option apply to partial content downloads for large files?

**Comment by eyalbe4** [comment]
Yes @scottrus,
The retries are counted and managed for each partial download separately, so if a specific segment fails, the other segmenets are not interrupted.

To get a better sense of the issue, try increasing **--retries** to a larger number, say 30 and also **--split-count**. Let us know if statistically you see any Improvement after changing this. It'll be interesting to know.

The log section you sent does not include the retries log messages. I'll have to look into the code to see if those messages aren't logged as INFO. If they're not, I think we should add that.

**Comment by scottrus** [comment]
Thanks @eyalbe4. No, the retry message is not logged at INFO level. Our failures are intermittent, about 1-2 a week across hundreds of jfrog CLI commands. I suspect the base cause is short network blips but cannot prove it yet as none of our other infrastructure is catching the issue in a log file.

**Comment by scottrus** [comment]
Question: is there also back off mech for retires or do they fire as fast as possible? For example, if this is a small network blip, retry behavior could be more reliable if if there was an increase in wait time between retries. (eg, 0.1s for retry one, 0.15s for retry two, 0.2s for retry three, etc)

**Comment by DimaNevelev** [comment]
Hi @scottrus,
We found that our retries mechanism doesn't work in case of unexpected EOF.
You can see our [downloadFileRange](https://github.com/jfrog/jfrog-client-go/blob/8e84d9c0eb93434161c7ad78d668be574041d76f/httpclient/client.go#L499) function. We are performing the retries in the sendGetForFileDownload function, which relays on the header only.
If the body transfer fails we won't do a retry.
We'll add a fix for this in one of the future releases.

**Comment by scottrus** [comment]
Yup, I see it now. Thanks for the pointers @DimaNevelev! So I'm wondering if the easy fix is to the retry in DownloadFileConcurrently? I'm thinking there are two possible ways to solve this:

1. Only retry the chunk that failed
1. Scrap all the chunks if any fail and retry the entire file and all chunks

I think the 2nd is easier to code by wrapping downloadChunksConcurrently in a retry method. Of course this over all this means more work (bandwidth, disk I/O) and possibly less chance of success if the network is very flaky. 

Short term I would be for the quicker solution.

**Comment by scottrus** [comment]
Checking to see if there is an update or more info I can provide.

**Comment by bloudraak** [comment]
I have also experienced this error on a regular basis. In our case we're downloading lots of small binaries but together they are around 200MB - 600MB. 

Any ETA for the fix? It is extremely disruptive when your CI/CD pipeline depends on those files being downloaded.

**Comment by scottrus** [comment]
Confirming we are still seeing this fairly regularly at 2-5 times per week.

**Comment by bloudraak** [comment]
We're seeing this numerous times a day; in fact, sometimes it totally blocks us from deploying artifacts from Artifactory.

**Comment by eyalbe4** [comment]
@scottrus and @bloudraak, 
We should have a pull request with a fix really soon. We'll let you know as soon as it is submitted.

**Comment by barbelity** [comment]
@scottrus @bloudraak,
We have released CLI version 1.23.2, which includes a fix for the discussed EOF issues.
You can take a look at the [relevant commit](https://github.com/jfrog/jfrog-client-go/commit/187d4ed1e6f7a8aa9c83dc51f0324bd3434248d9) in jfrog-client-go.
As your feedback is very helpful, please let us know if it solves your issue.
